### ISSUE TITLE
 Add XIncludes to Assembly Schema

### ISSUE THREAD

**Issue by tomschr** [docbook/docbook#54]
(Copied from https://sourceforge.net/p/docbook/rfes/310/)
# Problem

Currently, with DocBook 5.1 CR3 the schema has an `assembly.rn{c,g}` companion. However, this file/schema **does not** have any XIncludes elements at all.
# Proposed Solution

For writers who want to split up their assembly file into different parts and xinclude it, it would be convenient to add the XInclude elements into the Assembly schema. The user could then select between an `assembly.rn{c,g}` or `assemblyxi.rn{c,g}` schema.

**Comment by LarryRowland** [comment]
Since the assembly mechanism depends on cross references between the sections in an assembly, could you please provide an example of how a project using xinclude would be structured?  It is not entirely clear exactly what components would be xincluded at what levels.

**Comment by tomschr** [comment]
@LarryRowland Basically, first level elements from `assembly`. If we should go "deeper" or not is a question to debate.

Let's assume you want to build an assembly file which contains lots of different resources, modules, or other components. Such a file is probably very big and hard to maintain. In such a case, it would probably make sense to split the assembly file into smaller parts and xinclude them.

For example:

```xml
<assembly version="5.1"
    xmlns="http://docbook.org/ns/docbook"
    xmlns:xi="http://www.w3.org/2001/XInclude">
  <xi:include href="resources.xml"/>
  <xi:include href="struct-a.xml"/>
  <xi:include href="struct-b.xml"/>
</assembly>
```

The `resources.xml` file would include a file which contain the `<resources>` definition. The `<structure>` definition would be inside the files `struct-a.xml` and `struct-b.xml`.

Speaking in terms of RELAX NG, a customization layer could look like this:

```rnc
namespace a = "http://relaxng.org/ns/compatibility/annotations/1.0"
namespace ctrl = "http://nwalsh.com/xmlns/schema-control/"
default namespace db = "http://docbook.org/ns/docbook"
namespace html = "http://www.w3.org/1999/xhtml"
namespace mml = "http://www.w3.org/1998/Math/MathML"
namespace rng = "http://relaxng.org/ns/structure/1.0"
namespace s = "http://purl.oclc.org/dsdl/schematron"
namespace svg = "http://www.w3.org/2000/svg"
namespace xlink = "http://www.w3.org/1999/xlink"
#
namespace xi = "http://www.w3.org/2001/XInclude"
namespace local = ""

# complete definitions of db.xi.include omitted
db.xi.include =
    ## An XInclude
    element xi:include { db.xi.include.attlist, db.xi.fallback? }

include "assembly.rnc"
{

  db.assembly =
    ## Defines the hierarchy and relationships for a collection of resources
    element assembly {
        db.assembly.attlist,
        db.assembly.info,
        (db.resources+ | db.xi.include*),
        (db.structure  | db.xi.include)*,
        (db.relationships | db.xi.include)*,
        (db.transforms | db.xi.include)?
        }
}
```

I'm not sure if that makes sense or if we should allow it. Probably we should think about, if we need more. For example, the `<structure>` element contains `<module>`s. Should we allow to xinclude them too?

I think, it makes it easier to create "building blocks" for assembly files.

Hope my idea is clearer now. 😉

**Comment by LarryRowland** [comment]
@tomschr Sorry to take so long to get back to you and I appreciate your clarification.

I am curious as to how the contents of the externally composed structure elements would be validated during authoring, since the structure elements (as well as the relationship elements) depend on IDREF type attributes that reference xml:id attributes on elements that are contained in resource sections.  Additionally, transform elements have xml:id attributes that are referenced by other IDREF type attributes from other elements .  I don't see how to author these sections independently without the appropriate referenced elements.

It turns out there is nothing to prevent a resource from pointing at an assembly, which means that a module could simply point at a resource that pulled in an entire assembly and it should be processed appropriately.  I could see using xi:include to pull in a set of resources to be used by structures and relationships (assuming your authoring tool supports referencing xml:id attributes from included content) and transforms, but I'm not sure how to author content from files that don't have the elements that need to be referenced by xml:id values.

**Comment by tomschr** [comment]
@LarryRowland 

No worries Larry, all fine. 😄

> I am curious as to how the contents of the externally composed structure elements would be validated during authoring...

The same as you would do with any content authored in DocBook: you first resolve any XIncludes and then validate it with the Assembly schema. Whatever tool you use (some Java, xmllint, oXygen, or anything else) is secondary as long as all XIncludes are resolved prior to validation.

>  Additionally, transform elements have xml:id attributes that are referenced by other IDREF type attributes from other elements.

True, but isn't it the same as with modular DocBook content too? I'm curious why do you think is it different from content written as an Assembly in comparison to content written in DocBook?

> It turns out there is nothing to prevent a resource from pointing at an assembly, which means that a module could simply point at a resource that pulled in an entire assembly and it should be processed appropriately. I could see using xi:include to pull in a set of resources to be used by structures and relationships (assuming your authoring tool supports referencing xml:id attributes from included content) and transforms, but I'm not sure how to author content from files that don't have the elements that need to be referenced by xml:id values.

I'm not sure if I fully understand your concerns. Maybe I have overlooked something or being too naive. 😉 Maybe you could give an example about your concern.

From my point of view, writing a modular assembly file is not any different from writing a modular DocBook file (assuming modular means in this context files which are xincluded). Of course you have to make sure your IDs are correctly set and everything is valid. However, I see this more of a tools issue rather than a schema issue. The tools should deal with modular assembly files to resolve any XIncludes and report any ID problems.

**Comment by LarryRowland** [comment]
Thanks for the work on this, Tom.  I agree that minimizing the size of files that authors have to deal with is a reasonable goal (I hate to have time to go get a cup of coffee while a flat-file containing XML is loaded into an editor).

My question about validation was not on how to validate the file created by inclusion of content, but rather how content would be validated during authoring in the xincluded sections referenced here:
```
  <xi:include href="struct-a.xml"/>
  <xi:include href="struct-b.xml"/>
```
How would an author validate the content of the `struct-a.xml` and `struct-b.xml` files?  Wouldn't they have to include `resources` elements in each of the files, which would mean there would be duplicate `xml:id` attributes included in the file if the resources element referenced above them is used in all the files.

Sorry if I'm missing something, but it is not clear to me how validation would happen within the component files being included while they are being authored.  It seems that including complete `assembly` elements (perhaps isolating the xml:id attributes by namespaces automatically) would be a more reasonable approach than inclusion of elements that have to include cross-referencing `IDREF` and `xml:id` attributes.  I find validation a very helpful tool (including the pop-up lists of valid targets from cross references) at the authoring level, not just at the submission of content level.

**Comment by tomschr** [comment]
Hi Larry,

> How would an author validate the content of the `struct-a.xml` and `struct-b.xml` files? 

I'd say the `structure` elements and the `resource` elements can be independent. Let's look at this example:

```xml
<!-- assembly.xml -->
<assembly xmlns="http://docbook.org/ns/docbook"
          xmlns:xi="http://www.w3.org/2001/XInclude">
 <resources>
   <!-- ... omitted, but assume some resources are definied... -->
 </resources>

 <xi:include href="struct-a.xml"/>
 <xi:include href="struct-b.xml"/>
</assembly>
```

The above `assembly.xml` file defines some resources that would be used in the `struct-a.xml` and `struct-b.xml` files. These files have the root element `structure`, of course.

Validating the `struct-*.xml` files would be to validate the `structure` element. Sure, you need to take care of validating the ID/IDREF connection as well.

However, this is not something different than it would be a DocBook `section` element which is xincluded in an article. Usually, this is handled by your editor. For example, oXygen can do that with a "master file". In that sense, the `assembly.xml` file would be the master file and everything which is xincluded are some child files.

> Wouldn't they have to include resources elements in each of the files, which would mean there would be duplicate xml:id attributes included in the file if the resources element referenced above them is used in all the files.

I'd say no. :) With my above example, there is no need to duplicate any `resource` elements and as such, there are no duplicate `xml:id` attributes.

Does it help?

**Comment by LarryRowland** [comment]
@tomschr Thanks, Tom.  Apparently the confusion is mine.  I'm a heavy oxygen user, but have been using it for writing and debugging transforms for the last couple of years rather then authoring content (my job responsibilities have moved away from writing to infrastructure support).  It seems perfectly reasonable to provide the xi:include support you requested and I will make that clear at the next TC meeting.  It will, however, be important for tool developers to include support for interpreting the structures presented by the assembly in addition to the basic reference model presented by the actual xi:includes.  I'll check into it.

As far as where to add it to the schema beyond children of the root element, I would suspect the relatively flat nature of the resources element content would make it less valuable to allow importing the content of a resources element.  For module, since it can be a simple reference to a resource, I would not include xi:include on it.  The output elements can be fairly complex in terms of attribute combinations so might be reasonable candidates for inclusion.  I'm not sure how much info elements are likely to be included or have content that is included (although the fact that a module that is a child of a module could point at a resource that was an info element provides another route at inclusion of an info element), but it is probably reasonable to just use whatever inclusion attributes apply to the inclusion schema for DocBook elements to all DocBook elements included in the schema.  I think it would be less confusing that way.

Inside relationships, a relationship can be fairly complex, although I am not sure whether an individual relationship would be cherry-picked  to be part of a collection of relationships, it might happen.  I suspect that an individual instance would be too simple to be worth including directly as a component, and I'm not sure there would be much payoff for inclusion of the children of a relationship.

Given that transform elements are essentially elements to bind an external HREF to an xml:id so that they can be referenced by other elements in the assembly, I'm not sure they are good candidates for inclusion beyond pulling in a set of them, but since only a single transforms element is permitted, allowing for either multiple transforms elements or inclusion of individual transform elements should be considered.

This is a first stab at what might need to be modified.

**Comment by tomschr** [comment]
@LarryRowland Thanks for your patience and your help in bringing XIncludes to Assembly. 🎆 

> [...] As far as where to add it to the schema beyond children of the root element, I would suspect the relatively flat nature of the resources element content would make it less valuable to allow importing the content of a resources element. 

Depends. 😉 During testing some assembly files, I've found a discrepancy between the documented structure in TDG and the "real" assembly file found in http://docbook.org/xml/5.2b02/. I've opened issue #93, maybe you want to have a look? I suspect, this could influence this issue. At least it would be good to clarify what is authoritative.

Depending on what is allowed inside `resource` elements (only `description` or more DocBook elements?) it could be interesting to allow an additional XInclude.

---
>  For module, since it can be a simple reference to a resource, I would not include xi:include on it.

Do you mean the element `module` inside `structure` or just the generic term? IMHO, the `module` element can be fairly complex so I would see it as a natural candidate for replacing it with XIncludes.

For the other I agree with your careful considerations.

---

Ok, if I got your last comment right, I've summarized our findings into a checklist:

* [ ] Allow XIncludes at the first level after `assembly`; basically the customization layer from https://github.com/docbook/docbook/issues/54#issuecomment-302237102
* [ ] Inside `resources`? We need to check issue #93 first.
* [ ] Use XInclude as a replaceable for `info` elements?
* [ ] Add XIncludes as child to `relationships`?
* [ ] Probably don't allow XIncludes as a replaceable for `transform` elements.

Anything else? Did I miss anything or got it wrong from your comment? Maybe this can be used as a basis for discussion inside the committee.

Thanks for this fruitful discussion, I'm looking forward to whatever the committee will decide. 👍

**Comment by LarryRowland** [comment]
>> [...] As far as where to add it to the schema beyond children of the root element, I would suspect the relatively flat nature of the resources element content would make it less valuable to allow importing the content of a resources element.
> 
> Depends. 😉 During testing some assembly files, I've found a discrepancy between the documented structure in TDG and the "real" assembly file found in http://docbook.org/xml/5.2b02/. I've opened issue #93, maybe you want to have a look? I suspect, this could influence this issue. At least it would be good to clarify what is authoritative.
> 
> Depending on what is allowed inside resource elements (only description or more DocBook elements?) it could be interesting to allow an additional XInclude.
>
The documentation is in error.  There was a lot of debate about what should be allowed inside a resource element, One of the early models allowed any DocBook content inside a `resource` element and a different iteration allowed for referencing tools like `toc` and `index`, but the final decision was that only a description would be allowed, to provide for generation of manifests that had descriptions of available resources to help with authoring.  The more complex resource elements became, the more difficult it was to explain things to the users.  This is an old section that was not pulled out when the final schema was generated.
>>     For module, since it can be a simple reference to a resource, I would not include xi:include on it.
> 
> Do you mean the element module inside structure or just the generic term? IMHO, the module element can be fairly complex so I would see it as a natural candidate for replacing it with XIncludes.
> 
Should have been more careful about my markup 😁 :.  I was talking about `module` elements. If a `resource` element references a `module` element, it can be referenced by `module` elements with an `elementref` attribute.  When the `href` attribute includes #, it is implementation dependent what to do and the intent was to allow it to extract an element with content to provide extraction of element content from a file.  Thus:
```XML
...
<resource xml:id=" StockOverwrittenIntro" href="my_stock_modules#my_intro_overwrite">
    <description>This is the stock overwrite for the intro paragraphs.</description>
</resource>
...
<module resourceref="StockOverwrittenIntro"/>
...
```
would provide the same effect as a `module` element with an `xi:include` attribute.  There was discussion about whether using an implementation dependent # shortcut in the href instead of adding an optional attribute for some sort of XPATH value, however, the # shortcut is fairly well understood.

Given that the # shortcut is implementation dependent, it probably does make sense to permit `xi:include` on `module` elements.

> For the other I agree with your careful considerations.
> 
> Ok, if I got your last comment right, I've summarized our findings into a checklist:
> 
>     [x] Allow XIncludes at the first level after assembly; basically the customization layer from #54 (comment)
>     [ ] Inside resources? We need to check issue #93 first.
>     [x] Use XInclude as a replaceable for info elements?
>     [x] Add XIncludes as child to relationships?
>     [x] Probably don't allow XIncludes as a replaceable for transform elements.
> 
> Anything else? Did I miss anything or got it wrong from your comment? Maybe this can be used as a basis for discussion inside the committee.

Given that resource elements are actually incorrectly documented and should only permit `description` elements, I was leaning towards `xi:include` not being needed there, but I'm actually open to discussion on it, since I could see a standard `resource`, including the description, being included in multiple `resources` collections (such as a corporate overview or license description).

I am definitely planning on using this as a starting point for the discussions since this is a fairly complex area.

**Comment by tomschr** [comment]
> > issue #93 [...]
>
> The documentation is in error. There was a lot of debate about what should be allowed inside a resource element, [...]

Ok, I see. Thanks for your explanation, I assumed this already.

>  > Do you mean the element module inside structure or just the generic term? IMHO, the module element can be fairly complex so I would see it as a natural candidate for replacing it with XIncludes.
>
> Should have been more careful about my markup 😁 :. I was talking about module elements.

No worries. 😄 As a non-native English speaker, it's sometimes a challenge to get all things right. 😆

> If a resource element references a module element, it can be referenced by module elements with an elementref attribute

I only see a `resourceref` attribute, no `elementref`. Typo? :wink:

**Comment by ndw** [comment]
I'll see if I can get #93 fixed.

**Comment by LarryRowland** [comment]
@tomschr You called it -- typo.  I have fixed it.  One of the problems with interleaving this stuff between planning meetings for our next sprint is that sometimes when  I come back I end up thinking "What was I saying?"

@ndw Thanks for tracking these discussions.

**Comment by tomschr** [comment]
@LarryRowland Thanks, I know what you mean. 😄

@ndw Thanks for fixing #93. Much appreciated! 👍

**Comment by tomschr** [comment]
@LarryRowland Just to get the picture: I guess, the assembly schema is released together with the DocBook schema, right? I assume it will be in 5.2? Are there any specific dates for the release schedule? :wink:  Thanks! :+1:

**Comment by ndw** [comment]
I believe this has been implemented in DocBook 5.2.
